win32ss/user/ntuser/ghost.c

```diff
@@ -176,20 +443,64 @@ BOOL FASTCALL IntMakeHungWindowGhosted(HWND hwndHung)
 
     if (IntIsGhostWindow(pHungWnd))
     {
-        DPRINT1("IntIsGhostWindow\n");
+        ERR("IntIsGhostWindow\n");
         return FALSE;   // ghost window cannot be ghosted
     }
 
-    if (IntGhostWindowFromHungWindow(pHungWnd))
+    if (IntIsShellTrayWnd(pHungWnd))
     {
-        DPRINT("Already ghosting\n");
-        return FALSE;   // already ghosting
+        TRACE("IntIsShellTrayWnd\n");
+        return FALSE;   // Shell_TrayWnd
     }
 
-    // TODO: Find a way to pass the hwnd of pHungWnd to the ghost thread as we can't pass parameters directly
+    if (IntIsProgmanWnd(pHungWnd))
+    {
+        TRACE("IntIsProgmanWnd\n");
+        return FALSE;   // Progman
+    }
 
+    // initialize event
     if (!gptiGhostThread)
-        UserCreateSystemThread(ST_GHOST_THREAD);
+    {
+        InitializeListHead(&gGhostListHead);
+
+        // create event
+        gpGhostStartupEvent = ExAllocatePoolWithTag(NonPagedPool,
+                                                    sizeof(KEVENT),
+                                                    USERTAG_GHOST);
+        if (!gpGhostStartupEvent)
+        {
+            ERR("Unable to allocate event\n");
+            return FALSE;   // creation failed
+        }
+        KeInitializeEvent(gpGhostStartupEvent, SynchronizationEvent, FALSE);
+
+        // create ghost thread
+        if (!UserCreateSystemThread(ST_GHOST_THREAD))
+        {
+            ERR("Unable to create ghost thread for pHungWnd(%p), hwndHung(%p)\n",
+                (void *)pHungWnd, (void *)hwndHung);
+            return FALSE;   // creation failed
+        }
+    }
+
+    // wait for event
+    UserLeave();
+    Status = KeWaitForSingleObject(gpGhostStartupEvent, UserRequest, KernelMode, FALSE, NULL);
```

Comment:
Any reason to switch this to KernelMode? I think UserMode was appropriate.
But in case this stays KernelMode, the STATUS_USER_APC case can't happen, since a user APC wouldn't interrupt the wait.